src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AbstractAddMissingImportsFeatureService.cs

```diff
@@ -0,0 +1,150 @@
+                    }
+                }
+            }
+
+            return fixes.ToImmutableAndFree();
+        }
+
+        private async Task<Document> ApplyFixesAsync(Document document, ImmutableArray<AddImportFixData> fixes, CancellationToken cancellationToken)
+        {
+            if (fixes.IsEmpty)
+            {
+                return document;
+            }
+
+            var solution = document.Project.Solution;
+            var packageInstallerService = solution.Workspace.Services.GetService<IPackageInstallerService>();
+            var addImportService = document.GetLanguageService<IAddImportFeatureService>();
+
+            // Do not limit the results since we plan to fix all the reported issues.
+            var codeActions = addImportService.GetCodeActionsForFixes(document, fixes, packageInstallerService, limitResults: false);
+            var getChangesTasks = codeActions.Select(action => GetChangesForCodeActionAsync(document, action, cancellationToken));
+
+            await Task.WhenAll(getChangesTasks).ConfigureAwait(false);
+
+            // Using Sets allows us to accumulate only the distinct changes.
+            var allTextChanges = new HashSet<TextChange>();
+            // Some fixes require adding missing references.
+            var allAddedProjectReferences = new HashSet<ProjectReference>();
+            var allAddedMetaDataReferences = new HashSet<MetadataReference>();
+
+            foreach (var getChangesTask in getChangesTasks)
+            {
+                var (projectChanges, textChanges) = await getChangesTask.ConfigureAwait(false);
+
+                allTextChanges.UnionWith(textChanges);
+                allAddedProjectReferences.UnionWith(projectChanges.GetAddedProjectReferences());
+                allAddedMetaDataReferences.UnionWith(projectChanges.GetAddedMetadataReferences());
+            }
+
+            // Apply changes to both the project and document.
+            var newProject = document.Project.AddMetadataReferences(allAddedMetaDataReferences);
+            newProject = newProject.AddProjectReferences(allAddedProjectReferences);
+
+            var text = await document.GetTextAsync(cancellationToken);
+            var newText = text.WithChanges(allTextChanges);
+            var newDocument = newProject.GetDocument(document.Id).WithText(newText);
+
+            return newDocument;
+        }
+
+        private async Task<(ProjectChanges projectChanges, IEnumerable<TextChange> textChanges)> GetChangesForCodeActionAsync(Document document, CodeAction codeAction, CancellationToken cancellationToken)
+        {
+            var newSolution = await codeAction.GetChangedSolutionInternalAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
+            var newDocument = newSolution.GetDocument(document.Id);
+
+            var textChanges = await newDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false);
```

Comment:
on the fence about thsi... 

We need to do some testing of some significant pastes.  Here's why:  GetTextChanges is not guaranteed to offer any particular behavior.  It may sometimes say that the entire file changed, or it may do a bad diff that ends up with some crazy changes.  There are issues tracking this, but i can't find them.

So, basically, you're calling this assuming you'll get a nice minimal edit back stating which import was added.  But i don't know if you can trust that will happen.  And then, if you have weird stuff (like overlapping changes) i have no idea what blows up later.  This is a definite concern for me.

Reply:
Oh, here's one of hte bugs: https://github.com/dotnet/roslyn/issues/17498

Reply:
Note: even if we test this, i'm still very wary about this approach.  I think i'd much prefer to have the actual add-import service able to add everything.  rather than getting a bunch of codeactions back, and attempting to figure out how to reconcile all the work they're doing.  

I know this is likely *not* what you wanted to hear :-/

Reply:
@CyrusNajmabadi  In the earliest versions of this Service I would serially take each FixData and request a code action for the current state of the document, call GetGetChangedSolution, and update the document. This would avoid all of these issues for the time being correct? It might even make separate import groups work since requesting the changes against the unchanged document makes each change think it is first. =)

Reply:
That *might* work *if* FixData instances are idempotent and can be used in a different solution than the one they were created for.  I haven't checked. Are they actually suitable for that?

Reply:
Actuall, if you have all the FixData's... can you not just apply them all at the same time?

Reply:
Like... instead of creating an individual code action per fix data... just collect all the fix datas, and apply that as one big change.  Woudl that not work?